src/pubsub/ua_pubsub.c

```diff
@@ -182,6 +185,1098 @@ UA_Server_removeWriterGroup(UA_Server *server, const UA_NodeId writerGroup){
+        UA_QualifiedName qn;
+        UA_QualifiedName_init(&qn);
+        char szTmpName[64];
+        if (pDataSetReader->config.dataSetMetaData.fields[iteratorField].name.length > 0) {
+            int slen = 63;
+            vAttr.displayName.locale = UA_STRING("en-US");
+            vAttr.displayName.text = pDataSetReader->config.dataSetMetaData.fields[iteratorField].name;
+            if (pDataSetReader->config.dataSetMetaData.fields[iteratorField].name.length < 63) {
+                slen = (int)pDataSetReader->config.dataSetMetaData.fields[iteratorField].name.length;
+                UA_snprintf(szTmpName, sizeof(szTmpName), "%s", (const char*)pDataSetReader->config.dataSetMetaData.fields[iteratorField].name.data);
+            }
+
+            szTmpName[slen] = '\0';
+            qn = UA_QUALIFIEDNAME(1, szTmpName);
+        }
+        else {
+            strcpy(szTmpName, "SubscribedVariable");
+            vAttr.displayName = UA_LOCALIZEDTEXT("en-US", szTmpName);
+            qn = UA_QUALIFIEDNAME(1, "SubscribedVariable");
+        }
+
+        //Add variable to the given parent node
+        UA_NodeId newNode;
+        retVal = UA_Server_addVariableNode(server, UA_NODEID_NULL, *parentNode,
+                                       UA_NODEID_NUMERIC(0, UA_NS0ID_HASCOMPONENT), qn,
+                                       UA_NODEID_NUMERIC(0, UA_NS0ID_BASEDATAVARIABLETYPE), vAttr, NULL, &newNode);
+        if (retVal == UA_STATUSCODE_GOOD) {
+            UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_USERLAND, "addVariableNode %s succeeded", szTmpName);
+        }
+        else {
+            retval = retVal;
+            UA_LOG_WARNING(UA_Log_Stdout, UA_LOGCATEGORY_USERLAND, "addVariableNode: error 0x%x", retVal);
+        }
+
+        UA_FieldTargetDataType_init(&targetVars.targetVariables[iteratorField]);
+        targetVars.targetVariables[iteratorField].attributeId = UA_ATTRIBUTEID_VALUE;
+        UA_NodeId_copy(&newNode, &targetVars.targetVariables[iteratorField].targetNodeId);
+        UA_NodeId_deleteMembers(&newNode);
+        if (vAttr.arrayDimensionsSize > 0) {
+            UA_Variant_deleteMembers(&vAttr.value);
+            UA_Array_delete(vAttr.arrayDimensions, vAttr.arrayDimensionsSize, &UA_TYPES[UA_TYPES_UINT32]);
+        }
+    }
+
+    if (sdsType == UA_PUBSUB_SDS_TARGET) {
+        retval = UA_PubSubDataSetReader_createTargetVariables(server, pDataSetReader->identifier, &targetVars);
+    }
+
+    UA_TargetVariablesDataType_deleteMembers(&targetVars);
+    return retval;
+}
+
+/**
+ * Process a NetworkMessage with a DataSetReader.
+ *
+ * @param server
+ * @param dataSetReader
+ * @param dataSetMsg
+ */
+void UA_PubSubDataSetReader_process(UA_Server *server, UA_PubSubDataSetReader *dataSetReader, UA_DataSetMessage* dataSetMsg) {
+    if ((dataSetReader == NULL) || (dataSetMsg == NULL) || (server == NULL)) {
+        return;
+    }
+
+    if (dataSetMsg->header.dataSetMessageValid)
```

Comment:
Return early if the message is not valid --> saves one level of indendation of the code.